pkgs/development/tools/misc/cquery/default.nix

```diff
@@ -0,0 +1,54 @@
+{ stdenv, fetchFromGitHub, makeWrapper
+, cmake, llvmPackages, ncurses }:
+
+let
+  src = fetchFromGitHub {
+    owner = "cquery-project";
+    repo = "cquery";
+    rev = "e45a9ebbb6d8bfaf8bf1a3135b6faa910afea37e";
+    sha256 = "049gkqbamq4r2nz9yjcwq369zrmwrikzbhfza2x2vndqzaavq5yg";
+    fetchSubmodules = true;
+  };
+
+  stdenv = llvmPackages.stdenv;
+
+in
+stdenv.mkDerivation rec {
+  name    = "cquery-${version}";
+  version = "2018-03-25";
+
+  inherit src;
+
+  nativeBuildInputs = [ cmake makeWrapper ];
+  buildInputs = with llvmPackages; [ clang clang-unwrapped llvm ncurses ];
+
+  cmakeFlags = [
+    "-DSYSTEM_CLANG=ON"
+    "-DCLANG_CXX=ON"
+  ];
+
+  postFixup = ''
+    # We need to tell cquery where to find the standard library headers.
+
+    args="\"-isystem\", \"${if (stdenv.hostPlatform.libc == "glibc") then stdenv.cc.libc.dev else stdenv.cc.libc}/include\""
+    args+=", \"-isystem\", \"${llvmPackages.libcxx}/include/c++/v1\""
+
+    wrapProgram $out/bin/cquery \
+      --add-flags "'"'--init={"extraClangArguments": ['"''${args}"']}'"'"
```

Comment:
It gets overridden if it is duplicated, but probably test best we can do without custom patches. `NIX_CFLAGS_COMPILE`/`$NIX_CXXSTDLIB_COMPILE` might be also interesting in future.

Reply:
Sorry I did not test for that. I think it would be best to patch that upstream. I could use jd to merge the options in the meantime, but that would be pretty hacky...

NIX_CFLAGS_COMPILE contains thing that are only needed during build time, so I don't think that should be used.
See my other reply regarding NIX_CXXSTDLIB_COMPILE.

Btw, thank you for taking a look.

Reply:
I mean `NIX_CFLAGS_COMPILE` could be read at run-time because then developing in a nix shell would just works.

Reply:
I see. I'd still like it to be usable outside of nix-shell, though.
Also, reading NIX_CFLAGS_COMPILE at runtime would introduce platform dependent behaviour, and I would rather not provide people with an excuse to avoid creating a compile_commands.json file.

Reply:
The point is that on nix `compile_command.json` is not enough to find all includes.
To give you an example. If you install `zlib` on ubuntu it will be installed into `/usr/include`.
gcc/libclang will search there for includes and find it during compilation.
If you do `nix-shell -p zlib` in a project, cc-wrapper from `nixpkgs` will add `-isystem /nix/store/spxwrxnsf28sq5gdp42xvqac7psvlqya-zlib-1.2.11-dev/include -isystem /nix/store/spxwrxnsf28sq5gdp42xvqac7psvlqya-zlib-1.2.11-dev/include` to `$NIX_CFLAGS_COMPILE` and the compiler wrapper around gcc/clang will pick the same variable up.
In both versions the build system don't need to explicitly pass an include argument and they don't do that. So if `cquery` does not replicate the same behaviour it will be not very useful except for basic c/c++ project depending on the standard library.

Reply:
I would argue that if a ```compile_commands.json``` does not contain include paths to zlib if its header is included in your compilation unit, then the compilation database is wrong.
Even though the path to ```/usr/include``` is added automatically on other systems does not make it right. 
There are better ways to do this, for example cmake solves this through ```CMAKE_MODULES_PATH``` and the ```findXXX.cmake``` modules. 
I must admit that classic ```./configure; make; make install``` style build systems can't easily do that, and getting those dependencies from the ```NIX_CFLAGS_COMPILE``` field would be a viable workaround.

Reply:
But no real-life project does this correct completely. Take a look at neovim, systemd or llvm - these where just the first random three project I checked. Usually most projects just check if a header file can be found, but do not bother to add the full path - this would require either a cmake module or a pkgconfig file. Not every library/distribution provides this, sometimes they are even broken, that's why many projects add fallbacks to their build system.